src/crimson/osd/ops_executer.cc

```diff
@@ -83,10 +86,12 @@ OpsExecuter::call_errorator::future<> OpsExecuter::do_op_call(OSDOp& osd_op)
       return std::make_pair(ret, std::move(outdata));
     }
   ).then(
-    [prev_rd = num_read, prev_wr = num_write, this, &osd_op, flags]
+    [this, prev_rd, prev_wr, &osd_op, flags]
```

Comment:
why cannot we capture `num_read` and `num_write` using `[prev_rd = num_read, prev_wr = num_write, ...]`? what's the difference the revised code and the buggy code?

Reply:
Yeah, it's not intuitive. I think the binding really happens after the `method->exec()` call, so the situation is quite similar to what we have in the following snippet:

```cpp
template <class F>
void execute_now (F&& f) {
  f();
}

int main() {
    int val = 42;

    execute_now([&val] { val *= 2; });
    return [val_copy = val] { return val_copy; }();
}
```
```
$ /tmp/x; echo $?
84
```

Reply:
i see. in this case,
```c++
async([this] { this->val = NEW_VAL; }).then([old_val=val, this] { /**/ });
```

per http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0145r3.pdf

> The postfix-expression is sequenced before each expression in the expression-list and any default argument.

where `async([this] { this->val = NEW_VAL; }).then` is a postfix expression which selects "then" member function, and `[old_val=val, this] { /**/ }` is the expression-list. `async([this] { this->val = NEW_VAL; })` is yet another postfix expression. so `async([this] { this->val = NEW_VAL; }).then` is evaluated first.  because in this case, `async()` returns an "immediate" future, which is fulfilled right away, `this->val` is changed before it is captured when evaluating the following expression-list.